### ISSUE TITLE
Tutorial code for SGPR: training too long -> high training error

### ISSUE THREAD

**Issue by mkc1000** [cornellius-gp/gpytorch#2302]
I ran the tutorial code basically exactly*, but I changed 50 steps of optimization to 2000, and both the MAE and NLL increase significantly, not just for the test data, but for the training data! This happens even as the training loss reported by the torch optimizer goes steadily down. How is this possible, especially for NLL on the training data? There must be some bug, right?

Code is here: https://colab.research.google.com/drive/1-Fe53FnF5DWmZV4ZTbZrOFEZDESyMsXK?usp=sharing.

\* 1) I changed the inducing points to be a *clone* of some of the data points, so that when they are optimized, the original training data isn't changed. 2) I had to host the data file myself, and change the link, because the existing link sends you to a "can't scan for viruses" warning page. 3) I added NLL evaluation, in addition to MAE.

**Comment by gpleiss** [comment]
I cannot reproduce your error.

*100 training iterations:*
- Train loss: 0.362
- Test MAE: 0.0726
- Test NLL: 0.346

*2000 training iterations:*
- Train loss: -0.901
- Test MAE: 0.0722
- Test NLL: -0.795

A few notes:

1. Cloning the training data is a good catch. I'll update the notebook.
2. Your NLL computation is incorrect. You shouldn't be exponentiating the noise (it typically uses the soft plus transform). The correct way to compute this in GPyTorch is 

```python
preds = model.likelihood(model(test_x))  # gets predictive posterior, rather than posterior
nll = -preds.to_data_independent_dist().log_prob(test_y)
```

I'm updating the notebook to include these changes, but otherwise I'm not convinced there's any bug.

**Comment by mkc1000** [comment]
Hi Geoff, thanks so much for looking into this. I'm a bit mystified though, and not sure what I've done wrong. But I tried using only a CPU, and I got much better numbers--still not as good as yours.

Here is the code again: https://colab.research.google.com/drive/1-Fe53FnF5DWmZV4ZTbZrOFEZDESyMsXK?usp=sharing.

On a GPU, for training_iterations = 2000, I got:

Train Loss = -0.901
Test MAE: 0.2147902548313141
Test NLL: 26548.458984375
Train MAE: 0.2169969230890274
Train NLL: 26950.0390625

(I'm thinking the NLL code doesn't add the observation noise to the predictive distribution; is there a way to do that? Not such a problem, but would be nice).

On a CPU, for training_iterations = 2000, I got:

Train Loss = -0.901
Test MAE: 0.16272222995758057
Test NLL: 0.010032381862401962
Train MAE: 0.16091860830783844
Train NLL: -0.012847199104726315

Does this version info look right to you?

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Collecting gpytorch
  Downloading gpytorch-1.9.1-py3-none-any.whl (250 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.9/250.9 KB 4.9 MB/s eta 0:00:00
Requirement already satisfied: scikit-learn in /usr/local/lib/python3.9/dist-packages (from gpytorch) (1.2.2)
Collecting linear-operator>=0.2.0
  Downloading linear_operator-0.3.0-py3-none-any.whl (155 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 155.6/155.6 KB 12.7 MB/s eta 0:00:00
Requirement already satisfied: scipy in /usr/local/lib/python3.9/dist-packages (from linear-operator>=0.2.0->gpytorch) (1.10.1)
Requirement already satisfied: torch>=1.11 in /usr/local/lib/python3.9/dist-packages (from linear-operator>=0.2.0->gpytorch) (1.13.1+cu116)
Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.9/dist-packages (from scikit-learn->gpytorch) (3.1.0)
Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.9/dist-packages (from scikit-learn->gpytorch) (1.1.1)
Requirement already satisfied: numpy>=1.17.3 in /usr/local/lib/python3.9/dist-packages (from scikit-learn->gpytorch) (1.22.4)
Requirement already satisfied: typing-extensions in /usr/local/lib/python3.9/dist-packages (from torch>=1.11->linear-operator>=0.2.0->gpytorch) (4.5.0)
Installing collected packages: linear-operator, gpytorch
Successfully installed gpytorch-1.9.1 linear-operator-0.3.0

I also tried installing the latest (unstable) version--see the commented out pip commands--but got the same result; only tried this on the GPU.

**Comment by gpleiss** [comment]
I'm not sure what the issue is then... perhaps an issue with colab

**Comment by patel-zeel** [comment]
@mkc1000 I think the torch version plays the pivotal role here. I tried running it after installing several pytorch versions in colab, and the results are the following:

_2000 training iterations:_
_gpytorch version: 1.9.1_
| PyTorch version | Results |
| -----------------|---------|
| `1.11.0+cu102` | Train loss: -.901 <br> Test MAE:  0.0827 <br> Test NLL: 1391.853 |
| `1.12.1+cu102` | Train loss: -.901 <br> Test MAE: 0.0718 <br> Test NLL: -0.5459 |
| `1.13.1+cu116` (colab default) | Train loss: -.901 <br> Test MAE: 0.2147 <br> Test NLL: 26548.4628 |
| `2.0.0+cu117` | Train loss: -.901 <br> Test MAE:  0.0857 <br> Test NLL: 5659.292 |

@gpleiss This looks worrisome because results may change drastically if a different pytorch version is used. Have you experienced something similar with `torch` before? Would it help to recommend a specific pytorch version to gpytorch users?

Few more points:
* Given that `gpytorch.metrics` seems stabilized now, would it be useful to use them in tutorial notebooks? It may help people avoid wrongly calculating a metric.
* Dataset download link is broken due to a virus warning from Google drive. We may replace it with the following link (found by @mkc1000): 
    * [https://mkcohen-hosted-files.s3.us-west-1.amazonaws.com/elevators.mat](https://mkcohen-hosted-files.s3.us-west-1.amazonaws.com/elevators.mat)

**Comment by gpleiss** [comment]
@patel-zeel thanks for the thorough testing! I'll look into this more on Monday....

I think the `gpytorch.metrics` idea is a good idea for the notebooks. Also, switching the download link with the one you've posted.

**Comment by patel-zeel** [comment]
Thanks for the response and feedback, @gpleiss. 

@mkc1000 Would you like to fix the download link with a PR? I'll look into the `gpytorch.metrics` change.
